text/0013-api-calls-and-fhir.md

```diff
@@ -0,0 +1,76 @@
+# API Calls and FHIR
+- Start Date: 2019/07/02
+- RFC PR: https://github.com/openmrs/openmrs-rfc-frontend/pull/13
+
+## Decision, including impact on distributions
+All API calls from the browser to the server go through the `@openmrs/api` javascript
+module.
+
+The `@openmrs/api` provides the following:
+- `authenticatedFetch`
+- `fhir`
+- `getCurrentPatient`, and a few similar `getCurrent...` functions for shared data.
+
+Whenever a FHIR backend API is available, it should be used. When no such FHIR api exists,
+the OpenMRS-specific APIs will be used.
+
+## Definition
+- `fhir` refers to the [official fhir.js npm package](https://github.com/FHIR/fhir.js/). [FHIR itself](https://www.hl7.org/fhir/overview.html)
+  is an acronym that refers to the industry standard API specification for healthcare data. At the time of this writing, the OpenMRS
+  backend largely offers only non-FHIR APIs for the frontend to use.
+- Within a browser, an API call is an HTTP request to the server. This can be done with either
+  [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) or
+  [XMLHttpRequest XHR](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Using_XMLHttpRequest).
+  Fetch is preferred when possible, since it is newer and easier to work with (uses Promises).
+- Authentication of an API call is accomplished through a [`Cookie` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies)
+  or other [HTTP headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers). The `@openmrs/api` wrapper for fetch
+  adds any needed headers and handles server responses that indicate the user is not logged in.
+- `getCurrentPatient` and other `getCurrent...` functions are provided to share data between disparate microfrontends that
+  need the same data at the same time. They use the frontend route to identify the current patientUuid and then maintain a cache
+  for the shared object. The cache is busted whenever the url changes such that the patientUuid is no longer in the frontend url.
+
+## Reason for decision
+### `authenticatedFetch`
+
+Maintaining an OpenMRS-specific wrapper around `window.fetch()` allows us to make a single code change when
+improving or fixing authentication and other API behavior.
```

Comment:
@joeldenning  Not sure how this would work out, given that zone.js already has wrapped `window.fetch()`

Reply:
Since authenticated fetch calls window.fetch, zone.js will still be able to make the fetch promise a “Zone Aware promise” and trigger an Angular re-render when the fetch result comes back.

For angular applications, it might be best to inject our own HttpProvider that calls authenticated fetch instead of XHR or window.fetch. See “common practices section” where that is called out.